glotaran/builtin/models/kinetic_image/kinetic_image_dataset_descriptor.py

```diff
@@ -12,11 +12,38 @@
     }
 )
 class KineticImageDatasetDescriptor(DatasetDescriptor):
-    def get_k_matrices(self):
-        return [mat for mat in [cmplx.full_k_matrix() for cmplx in self.megacomplex] if mat]
+    def get_megacomplex_k_matrices(self):
+        scales = (
+            [
+                self.megacomplex_scale[i]
+                for i in [
+                    i
+                    for i, megacomplex in enumerate(self.megacomplex)
+                    if megacomplex.has_k_matrix()
+                ]
+            ]
```

Comment:
```suggestion
            [
                self.megacomplex_scale[i]
                    i
                    for i, megacomplex in enumerate(self.megacomplex)
                    if megacomplex.has_k_matrix()
            ]
```

This should do the same but with one less loop and without using `i` as a variable name twice.

Also since  we have two loops which check for `megacomplex.has_k_matrix()` we might want to create a filtered version and use that instead, e.g.:

```python
enumerated_megacomplices_with_k_matrix = [
    index, megacomplex
    for index, megacomplex in enumerate(self.megacomplex)
    if megacomplex.has_k_matrix()
]
```

Reply:
This has_kmatrix will be obsolete in future since I will revice megacomplex/model composition.